crates/turborepo-lib/src/task_graph/visitor.rs

```diff
@@ -577,22 +601,52 @@ impl ExecContext {
         };
 
         match result {
-            ExecOutcome::Success => {
+            ExecOutcome::Success(outcome) => {
+                let task_summary = match outcome {
+                    SuccessOutcome::CacheHit => tracker.cached().await,
+                    SuccessOutcome::Run => tracker.build_succeeded(0).await,
+                };
                 callback.send(Ok(())).ok();
+                if let Some(client) = spaces_client {
+                    let logs = logs.expect("spaces enabled logs should be collected");
+                    let info = self.spaces_task_info(self.task_id.clone(), task_summary, logs);
+                    client.finish_task(info).await.ok();
+                }
             }
             ExecOutcome::Internal => {
+                tracker.cancel();
                 callback.send(Err(StopExecution)).ok();
                 self.manager.stop().await;
             }
-            ExecOutcome::Task => {
+            ExecOutcome::Task { exit_code, message } => {
+                let task_summary = tracker.build_failed(exit_code, message).await;
                 callback
                     .send(match self.continue_on_error {
                         true => Ok(()),
                         false => Err(StopExecution),
                     })
                     .ok();
-                if !self.continue_on_error {
-                    self.manager.stop().await;
+
+                match (spaces_client, self.continue_on_error) {
```

Comment:
There is most certainly a better way to do this, but everything I tried ran into lifetime issues. I'll continue thinking about it, but I didn't want to block this PR on being clever.

Reply:
I don't see anything wrong with this? What's the issue?

Reply:
I just don't love explicitly handling ever case separately just to allow for us to concurrently run the two futures if there are two to run.